### PR TITLE
Fix name shortening

### PR DESCRIPTION
**Pull request by kamleshbhalui**
1. This PR has code extracted from table config PR related to name shortening
2. Fix JumpAction name

### CODE REVIEW

**Review by usha1830 — APPROVED** [review]
Looks fine

**Review by psivanup — APPROVED** [review]
Looks fine to me.

File: backends/dpdk/dpdkAsmOpt.h [hunk 1]
```diff
@@ -246,18 +246,65 @@ class ShortenTokenLength : public Transform {
...
 
+    const IR::Node* preorder(IR::ActionList* al) override {
+        IR::IndexedVector<IR::ActionListElement> new_al;
+        for (auto ale : al->actionList) {
+            auto methodCallExpr = ale->expression->to<IR::MethodCallExpression>();
+            auto pathExpr = methodCallExpr->method->to<IR::PathExpression>();
+            auto path0 = pathExpr->path->clone();
+            // this for getting rid of NoAction_0 or NoAction_1
+            if (path0->name.name.find("NoAction"))
+                path0 = new IR::Path("NoAction");
```
Comment (syogender) [inline backends/dpdk/dpdkAsmOpt.h:277]: we are leaking path0 here
you can search in pathExpr->path for "noaction" and clone only when it is noaction

File: backends/dpdk/dpdkAsmOpt.h [hunk 2]
```diff
@@ -246,18 +246,65 @@ class ShortenTokenLength : public Transform {
...
+        return new IR::ActionList(al->srcInfo, new_al);
+    }
+
     const IR::Node* preorder(IR::DpdkTable *t) override {
         t->name = shortenString(t->name);
+        auto methodCallExpr = t->default_action->to<IR::MethodCallExpression>();
+        auto pathExpr = methodCallExpr->method->to<IR::PathExpression>();
+        auto path0 = pathExpr->path->clone();
+        // this for getting rid of NoAction_0 or NoAction_1
+        if (path0->name.name.find("NoAction"))
```
Comment (syogender) [inline backends/dpdk/dpdkAsmOpt.h:297]: can we write a common function which can be called from this function and above function

**Review by syogender — COMMENTED** [review]
what is expected in spec file?
I thought for large name a short name will be used

**Comment by kamleshbhalui** [comment]
> what is expected in spec file? I thought for large name a short name will be used

thats extactly happening.
without this pr:
`jmpa LABEL_SWITCH vxlan_encap_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk`
after this pr
`jmpa LABEL_SWITCH vxlan_encap_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dpdk_dp5`

File: backends/dpdk/dpdkAsmOpt.h (hunk 1 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkAsmOpt.h]: fixed

File: backends/dpdk/dpdkAsmOpt.h (hunk 2 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkAsmOpt.h]: done

File: testdata/p4_16_psa_errors_outputs/psa-example-dpdk-directmeter-err.p4-error [hunk 3]
```diff
@@ -1,3 +1,3 @@
-psa.p4(638): [--Werror=unexpected] error: execute method of meter0 extern can only be invoked from within action of ownertable
+psa.p4(657): [--Werror=unexpected] error: execute method of meter0 extern can only be invoked from within action of ownertable
```
Comment (usha1830) [inline testdata/p4_16_psa_errors_outputs/psa-example-dpdk-directmeter-err.p4-error:1]: The method name should have been dpdk_execute

File: backends/dpdk/dpdkAsmOpt.h [hunk 4]
```diff
@@ -174,18 +174,23 @@ class ShortenTokenLength : public Transform {
     // 1.30.30 => 63(including dot(.))
     // if id name less than allowedLength keep it same
     cstring shortenString(cstring str, size_t allowedLength = 60) {
-        if (str.size() <= allowedLength)
+        // this for getting rid of suffixes in NoAction_0 or NoAction_1
+        if (str.size() > 8 && str.substr(0, 8) == "NoAction") {
```
Comment (usha1830) [inline backends/dpdk/dpdkAsmOpt.h:178]: This string manipulation can be avoided if you check originalName in caller. Can we do that?

File: testdata/p4_16_psa_errors_outputs/psa-meter6.p4-error [hunk 5]
```diff
@@ -1,6 +1,6 @@
 psa-meter6.p4(51): [--Wwarn=unsupported] warning: MyIC.meter0: Packet metering is not supported. Falling back to byte metering.
     DirectMeter(PSA_MeterType_t.PACKETS) meter0;
                                          ^^^^^^
-psa.p4(638): [--Werror=unexpected] error: execute method of meter0 extern can only be invoked from within action of ownertable
+psa.p4(657): [--Werror=unexpected] error: execute method of meter0 extern can only be invoked from within action of ownertable
```
Comment (usha1830) [inline testdata/p4_16_psa_errors_outputs/psa-meter6.p4-error:4]: This is non-dpdk test, so should be fine.

File: testdata/p4_16_psa_errors_outputs/psa-example-dpdk-directmeter-err.p4-error (hunk 3 above)
Comment (kamleshbhalui) [reply testdata/p4_16_psa_errors_outputs/psa-example-dpdk-directmeter-err.p4-error:1]: Can we get rid of hardcode method names from error messages?

File: backends/dpdk/dpdkAsmOpt.h (hunk 4 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkAsmOpt.h:178]: I'm not sure dpdk ir has that

Comment (usha1830) [reply backends/dpdk/dpdkAsmOpt.h:178]: (snip)
```
class DpdkAction {
    optional Annotations        annotations = Annotations::empty;
    inline IndexedVector<DpdkAsmStatement> statements;
    inline ID name;
...
}

``` 

   name.originalName should be available.

**Pull request closed without merging**